### PR TITLE
fix: fixed threshold for columns with units

### PR DESCRIPTION
**Pull request by SagarRajput-7**
### Summary

- Fixed threshold not working when comparing columns with units

Additionally:
- Added conversion of units in the same category like `ns` to `ms` 
- Restrict category select to only columnUnit 
- Restricted column name from threshold option

#### Related Issues / PR's

https://github.com/SigNoz/signoz/issues/5262

#### Screenshots

https://github.com/user-attachments/assets/da10615d-19eb-46a7-af85-70659eaeb977

#### Affected Areas and Manually Tested Areas

- Tested the threshold component under dashboard edit mode only, not used anywhere else

### CODE REVIEW

**Review by rahulkeswani101 — APPROVED** [review]

File: frontend/src/container/NewWidget/RightContainer/constants.ts [hunk 1]
```diff
@@ -11,10 +13,31 @@ export const operatorOptions: DefaultOptionType[] = [
...
+			options: getCategorySelectOptionByName(category),
+		}));
+	}
+	return categoryToSupport
+		.filter((supportedCategory) => supportedCategory === category)
+		.map((filteredCategory) => ({
+			label: filteredCategory,
+			options: getCategorySelectOptionByName(filteredCategory),
+		}));
+};
```
Comment (ahmadshaheer) [inline frontend/src/container/NewWidget/RightContainer/constants.ts:40]: nit: can you move these out of constants.ts?

**Review by ahmadshaheer — APPROVED** [review]

File: frontend/src/container/NewWidget/RightContainer/Threshold/Threshold.styles.scss [hunk 2]
```diff
@@ -297,6 +297,16 @@
 				box-shadow: none;
 			}
 		}
+
+		.invalid-unit {
+			color: var(--bg-vanilla-400);
+			font-family: 'Space Mono';
```
Comment (YounixM) [inline frontend/src/container/NewWidget/RightContainer/Threshold/Threshold.styles.scss:303]: We use Giest Mono now, Please update this accordingly.

**Comment by vikrantgupta25** [comment]
@SagarRajput-7 please add necessary code comments to the functions, params , returns and what is it doing as well. it's hard to follow up without that

File: frontend/src/container/NewWidget/RightContainer/Threshold/Threshold.styles.scss (hunk 2 above)
Comment (SagarRajput-7) [reply frontend/src/container/NewWidget/RightContainer/Threshold/Threshold.styles.scss]: Done

File: frontend/src/container/NewWidget/RightContainer/constants.ts (hunk 1 above)
Comment (SagarRajput-7) [reply frontend/src/container/NewWidget/RightContainer/constants.ts]: Done

**Comment by SagarRajput-7** [comment]
> @SagarRajput-7 please add necessary code comments to the functions, params , returns and what is it doing as well. it's hard to follow up without that

@vikrantgupta25 - added please check

File: frontend/src/container/GridTableComponent/index.tsx [hunk 3]
```diff
@@ -97,13 +97,19 @@ function GridTableComponent({
 
 	const newColumnData = columns.map((e) => ({
 		...e,
-		render: (text: string): ReactNode => {
-			const isNumber = !Number.isNaN(Number(text));
+		render: (text: string, ...rest: any): ReactNode => {
+			let textForThreshold = text;
+			if (columnUnits && columnUnits?.[e.title as string]) {
+				textForThreshold = rest[0][`${e.title}_without_unit`];
```
Comment (vikrantgupta25) [inline frontend/src/container/GridTableComponent/index.tsx:103]: any ? can we add some type here for the rest props if possible ? and can add here that `without_unit` column is added by client itself for other calculations so this naming needs to be consistent with that

File: frontend/src/container/GridTableComponent/utils.ts [hunk 4]
```diff
@@ -30,10 +31,39 @@ function evaluateCondition(
...
+ * @param thresholdUnit - The unit to which the value should be converted.
+ * @param columnUnits - The current unit of the value.
+ * @returns A boolean indicating whether the value meets the threshold condition.
+ */
+function evaluateThresholdWithConvertedValue(
+	value: number,
+	thresholdValue: number,
+	thresholdOperator?: string,
+	thresholdUnit?: string,
+	columnUnits?: string,
```
Comment (SagarRajput-7) [inline frontend/src/container/GridTableComponent/utils.ts:51]: change to columnUnit

File: frontend/src/container/NewWidget/RightContainer/dataFormatCategories.ts [hunk 5]
```diff
@@ -438,3 +438,169 @@ export const dataTypeCategories: DataTypeCategories = [
...
+// Function to convert a value from one unit to another
+export function convertUnit(
+	value: number,
+	fromUnitId?: string,
+	toUnitId?: string,
+): number | null {
+	// Finds the category that contains the specified units
+	const category = dataTypeCategories.find((category) =>
+		category.formats.some(
+			(format) => format.id === fromUnitId || format.id === toUnitId,
```
Comment (SagarRajput-7) [inline frontend/src/container/NewWidget/RightContainer/dataFormatCategories.ts:576]: check if both category can be done here, to save below 2 find comparisons

Comment (SagarRajput-7) [reply frontend/src/container/NewWidget/RightContainer/dataFormatCategories.ts]: clubbed in one function loop

File: frontend/src/container/GridTableComponent/index.tsx (hunk 3 above)
Comment (SagarRajput-7) [reply frontend/src/container/GridTableComponent/index.tsx:103]: assigning any type is tough here, as it changes mostly and comes from BE

**Review by vikrantgupta25 — APPROVED** [review]

**Review by YounixM — APPROVED** [review]
